### ISSUE TITLE
Scala project within multi module Grade build utilising 'includeFlat' will not build

### ISSUE THREAD

**Issue by gravelld** [scalacenter/bloop#1488]
Original discussion: https://gitter.im/scalacenter/bloop?at=607564bdae90f368407a336a

I'm trying to install and compile some Scala projects which are part of a Gradle multi module build. End result is I want to use Metals, but Metals isn't playing nice, so I thought trying to compile via bloop might be a good first step to work out what is happening.
I have installed bloop. However, when I run the compile I just get nothing:

    $ bloop compile com.elsten.google.gson
    $ 

I'm guessing this is a common issue but I can't find any docs about this...

```
$ cat .bloop/com.elsten.google.gson.json 
{
    "version": "1.4.0",
    "project": {
        "name": "com.elsten.google.gson",
        "directory": "/home/gravelld/git-repo/com.elsten.google.gson",
        "workspaceDir": "/home/gravelld/git-repo/elsten-software_bliss",
        "sources": [
            "/home/gravelld/git-repo/com.elsten.google.gson/src/main/java"
        ],
        "dependencies": [

        ],
        "classpath": [

        ],
        "out": "/home/gravelld/git-repo/elsten-software_bliss/.bloop/com.elsten.google.gson/build",
        "classesDir": "/home/gravelld/git-repo/elsten-software_bliss/.bloop/com.elsten.google.gson/build/classes",
        "resources": [
            "/home/gravelld/git-repo/com.elsten.google.gson/src/main/resources"
        ],
        "java": {
            "options": [
                "-source",
                "1.8",
                "-target",
                "1.8",
                "-h",
                "/home/gravelld/git-repo/com.elsten.google.gson/build/generated/sources/headers/java/main",
                "-g",
                "-sourcepath",
                "-proc:none",
                "-XDuseUnsharedTable=true"
            ]
        },
        "platform": {
            "name": "jvm",
            "config": {
                "home": "/usr/lib/jvm/adoptopenjdk-11-hotspot-amd64",
                "options": [

                ]
            },
            "mainClass": [

            ],
            "classpath": [

            ]
        },
        "tags": [
            "library"
        ]
    }
}
```

```
$ bloop compile --verbose com.elsten.google.gson
[D] Loading workspace settings from bloop.settings.json
[D] Computing sources and classpath hashes for com.elsten.google.gson
[D] Scheduling compilation for com.elsten.google.gson...
[D] Ignoring analysis for com.elsten.google.gson, directory /home/gravelld/git-repo/elsten-software_bliss/.bloop/com.elsten.google.gson/build/classes-empty-com.elsten.google.gson is missing
[D] Increasing counter for /home/gravelld/git-repo/elsten-software_bliss/.bloop/com.elsten.google.gson/build/classes-empty-com.elsten.google.gson to 5
[D] Recording new last successful request for com.elsten.google.gson associated with /home/gravelld/git-repo/elsten-software_bliss/.bloop/com.elsten.google.gson/build/classes-empty-com.elsten.google.gson
[D] Elapsed: 10.682663 ms
```

```
$ bloop compile --debug com.elsten.google.gson
Unrecognized log context: com.elsten.google.gson
```

After running compile, the classes folder doesn't exist:

    $ ls .bloop/com.elsten.google.gson/build/classes-empty-com.elsten.google.gson 
    ls: cannot access '.bloop/com.elsten.google.gson/build/classes-empty-com.elsten.google.gson': No such file or directory

The project I'm trying to compile is not dependent on any source projects, Java or otherwise:

```
apply plugin: 'scala'
apply plugin: 'maven-publish'

version = '0.1'

dependencies {

    compile libraries.scala_library
    compile libraries.gson

}
```

In the root project, we do add the java plugin to sub-projects:

```
allprojects {

    apply plugin: 'java'
    apply plugin: 'bloop'
        ...
```
So com.elsten.google.gson is both a Java, Scala and bloop project.

There are no classes at all inside the `.bloop` folder:

    $ find .bloop/ -name *.class
    $
    $ tree .bloop/com.elsten.google.gson/build/
    .bloop/com.elsten.google.gson/build/
    └── bloop-bsp-clients-classes
        ├── classes-bloop-cli
        └── classes-Metals-m_RT1-QXSXi0_gWM4JP_2g==

**Comment by gravelld** [comment]
Note that the source set shows `src/main/java`! Changing it to `src/main/scala` appears to make no difference.

Also, there are no `dependencies` listed, but the build does define JAR based dependencies (not source dependencies).

**Comment by Arthurm1** [comment]
Hi @gravelld 

What version of the Bloop gradle plugin are you using? 1.4.8?
What version of Gradle are you using? >5.0 <7.0?
What's the definition of `libraries.scala_library` "org.scala-lang:scala-library:2.13.5"?

The Gradle export seems to think this is a pure Java project and doesn't have the Scala plugin applied.

**Comment by gravelld** [comment]
Yep, 1.4.8, apologies for not posting this before.

Gradle 6.8.3.

	scala_library: 'org.scala-lang:scala-library:2.12.8',

So my next step was to move all configuration contributed by the root project into the subproject so with total control, see if I could just get this small project building.

TL;DR: **I did a bunch of frigging about on this, but to cut to what the fix appears to have been...**

What I used to have was the root project being called `master` at the same level as the other projects. This is a common Gradle pattern: https://stackoverflow.com/questions/24827146/how-to-include-projects-in-multiproject-build-from-master-directory-in-gradle

When you do that, in the `settings.gradle` you use `includeFlat` rather than `include` to include each subproject.

I'd moved the master project into the root to see if that was affecting things, but it was only after I changed `includeFlat` to `include` that things started working.

This is now generating class files as expected:

```
$ bloop compile com.elsten.google.gson
Compiling com.elsten.google.gson (2 Scala sources)
Compiled com.elsten.google.gson (797ms)
```

So now back to Metals!

**Comment by Arthurm1** [comment]
Hi @gravelld 

I'll look to add `includeFlat` support.  You shouldn't have to change your project to use Bloop/Metals

**Comment by gravelld** [comment]
@Arthurm1 I've re-opened this as you have implied this should be fixed - I also changed the title to reflect the issue.

**Comment by Arthurm1** [comment]
@gravelld 

Just to confirm I've got your `includeflat` project layout correct...
```
home
└──gravelld
    └── git-repo
        └── com.elsten.google.gson
            └── sourcefiles
            └── build.gradle
        └── elsten-software_bliss
            ├── build.gradle
            ├── gradlew
            └── settings.gradle
```

So you always have to execute `gradlew` commands from within the `elsten-software_bliss`  or `com.elsten.google.gson` dirs and you can't execute them when you're in the `git-repo` dir?

**Comment by gravelld** [comment]
No, it's like this:

```
home
└──gravelld
    └── git-repo
        └── com.elsten.google.gson
            └── sourcefiles
            └── build.gradle
        └── elsten-software_bliss
            ├── build.gradle
            ├── gradlew
            └── settings.gradle

**Comment by gravelld** [comment]
You can execute Gradle commands in any folder - specific tasks can be added to specific subprojects. In fact, my main build is a separate subproject.

If you're in the root `elsten-software_bliss` folder you can't really do anything even if you have a gradle executable in your PATH - it doesn't know to route your request to `master` and so it doesn't know what the subprojects are.
